python/private/repo_utils.bzl

```diff
@@ -262,12 +341,50 @@ def _outputs_to_str(result):
             lines.append("<{} empty>".format(name))
     return "\n".join(lines)
 
+def _get_platforms_os_name(rctx):
+    """Return the name in @platforms//os for the host os.
+
+    Args:
+        rctx: repository_ctx
+
+    Returns:
+        `str | None`. The target name if it maps to known platforms
+        value, otherwise None.
+    """
+    os = rctx.os.name.lower()
+    if "linux" in os:
+        return os
+    if "windows" in os:
+        return "windows"
+    if "mac" in os:
+        return "osx"
+
+    return None
+
+# TODO: Remove after Bazel 6 support dropped
+def _watch(rctx, *args, **kwargs):
+    """Calls rctx.watch, if available."""
+    if hasattr(rctx, "watch"):
+        rctx.watch(*args, **kwargs)
+
+# TODO: Remove after Bazel 6 support dropped
+def _watch_tree(rctx, *args, **kwargs):
+    """Calls rctx.watch_tree, if available."""
+    if hasattr(rctx, "watch_tree"):
+        rctx.watch_tree(*args, **kwargs)
+
 repo_utils = struct(
+    # keep sorted
+    debug_print = _debug_print,
     execute_checked = _execute_checked,
-    execute_unchecked = _execute_unchecked,
     execute_checked_stdout = _execute_checked_stdout,
+    execute_unchecked = _execute_unchecked,
+    get_platforms_os_name = _get_platforms_os_name,
+    getenv = _getenv,
     is_repo_debug_enabled = _is_repo_debug_enabled,
-    debug_print = _debug_print,
-    which_checked = _which_checked,
     logger = _logger,
+    watch = _watch,
+    watch_tree = _watch_tree,
+    which_checked = _which_checked,
+    which_unchecked = _which_unchecked,
 )
```

Comment:
It feels like this is getting big, but it's good to leave this as is for now. All of the things are low level `repository_ctx` related things, so it's fine.

Reply:
I like how go has the `ctx` and passes it around everywhere. It is kind of a sign that IO is being done here or there. That makes pure starlark functions really obvious.

So I like the current design a lot. :)